Update unit tests to use `AEPTestUtils`

timkimadobe opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description

This PR is a continuation of the review process in #42 and refactors existing unit tests to use the `AEPTestUtils` and JSON comparison methods with the following changes:

**MediaPublicTrackerTests.swift**
1. All custom `isEqual` comparisons updated to use JSON comparison instead of custom dictionary casting and equality

**MediaXDMEventTests**
* testToXDMData - updated to use JSON comparison with strict equality
* For property `xdm.mediaCollection.sessionDetails.length` Is standard Int conversion fine (via AnyCodable decoder) instead of specific conversion to Int64?

**MediaXDMEventHelperTests**
1. Replace AssertUtils.compareSizeAndKeys with JSON comparison assertEqual

**XDMAdvertisingDetailsTests
XDMAdvertisingPodDetailsTests
XDMErrorDetailsTests
XDMChapterDetailsTests
XDMSessionDetailsTests
XDMMediaCollectionTests**
* `testEncode` does not have strict collection count equality validation so assertExactMatch was used - is exact equality desired instead?
    * Please note that as long as expected is valid, `actual` will be validated accordingly so guard/unwrap of `actual` is not strictly needed, unless there is a specific step that should be validated in depth

## Questions for reviewers
1. For dictionary value comparisons using the JSON comparison methods, `Double` values dont use the delta comparison logic present in `MediaPublicTrackerTests` `isEqual` method. Is this an acceptable change?:

from Sources -> Double+Media.swift:
```swift
// Adds functionality to compare the equality of two doubles to within the specified delta.
extension Double {
    func isAlmostEqual(_ doubleToCompare: Double, accuracy: Double = 0.000001) -> Bool {
        return fabs(self - doubleToCompare) < accuracy
    }
}
```
## Related Issue

<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->

## Motivation and Context

<!--- Why is this change required? What problem does it solve? -->

## How Has This Been Tested?

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Screenshots (if appropriate):

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

timkimadobe committed:
Add pod dependency for AEPTestUtils

timkimadobe committed:
Remove unneeded test helpers

timkimadobe committed:
Refactor Media+PublicAPITests.swift to use AEPTestUtils

timkimadobe committed:
Refactor MediaPublicTrackerTests to use AEPTestUtils and JSON comparison

timkimadobe committed:
Refactor MediaTests to use AEPTestUtils

timkimadobe committed:
Refactor MediaXDMEventHelperTests to use AEPTestUtils and JSON comparison

timkimadobe committed:
Refactor unit tests to use AEPTestUtils and JSON comparison

timkimadobe committed:
Update pods

timkimadobe committed:
Remove local MockExtension in favor of AEPTestUtils version

timkimadobe committed:
Update MediaPublicAPITests to use AnyCodable asserts

timkimadobe committed:
Remove commented out code

timkimadobe committed:
Update AEPTestUtils dependency to 5.0.0

timkimadobe committed:
Remove refactored assertions

kevinlind commented on Tests/UnitTests/XDMAdvertisingDetailsTests.swift:
```diff
@@ -29,22 +30,29 @@ class XDMAdvertisingDetailsTests: XCTestCase {
+        }
+        """
+
+        assertExactMatch(expected: expected, actual: decodedAdDetails)
```
I'd suggest using `assertEqual` here to make the test more strict. If the class under test is updated in the future, this test will fail causing the developer to update the tests with the changes.

timkimadobe replied:
Updated to use `assertEqual` across the board!

kevinlind commented on Tests/UnitTests/XDMAdvertisingPodDetailsTests.swift:
```diff
@@ -11,23 +11,31 @@
+        }
+        """
+
+        assertExactMatch(expected: expected, actual: decodedAdBreakDetails)
```
Use `assertEqual` here.

kevinlind commented on Tests/UnitTests/XDMErrorDetailsTests.swift:
```diff
@@ -11,22 +11,30 @@
+        }
+        """
+
+        assertExactMatch(expected: expected, actual: decodedErrorDetails)
```
Use `assertEqual` here.

kevinlind commented on Tests/UnitTests/XDMMediaCollectionTests.swift:
```diff
@@ -44,60 +45,75 @@ class XDMMediaCollectionTests: XCTestCase {
+        }
+        """
+
+        assertExactMatch(expected: expected, actual: decodedMediaCollection)
```
Use `assertEqual`

kevinlind commented on Tests/UnitTests/XDMMediaCollectionTests.swift:
```diff
@@ -44,60 +45,75 @@ class XDMMediaCollectionTests: XCTestCase {
+        }
+        """
+
+        assertExactMatch(expected: expected, actual: decodedMediaCollection)
```
Use `assertEqual`

kevinlind commented on Tests/UnitTests/XDMMediaCollectionTests.swift:
```diff
@@ -112,42 +128,60 @@ class XDMMediaCollectionTests: XCTestCase {
+        }
+        """
+
+        assertExactMatch(expected: expected, actual: decodedMediaCollection)
```
Use `assertEqual`

kevinlind commented on Tests/UnitTests/XDMMediaCollectionTests.swift:
```diff
@@ -112,42 +128,60 @@ class XDMMediaCollectionTests: XCTestCase {
+        }
+        """
+
+        assertExactMatch(expected: expected, actual: decodedMediaCollection)
```
Use `assertEqual`

kevinlind commented on Tests/UnitTests/XDMMediaCollectionTests.swift:
```diff
@@ -164,14 +198,27 @@ class XDMMediaCollectionTests: XCTestCase {
+        }
+        """
+
+        assertExactMatch(expected: expected, actual: decodedMediaCollection)
```
Use `assertEqual`

kevinlind commented on Tests/UnitTests/XDMMediaCollectionTests.swift:
```diff
@@ -181,13 +228,26 @@ class XDMMediaCollectionTests: XCTestCase {
+        }
+        """
+
+        assertExactMatch(expected: expected, actual: decodedMediaCollection)
```
Use `assertEqual`

kevinlind commented on Tests/UnitTests/XDMSessionDetailsTests.swift:
```diff
@@ -43,45 +44,50 @@ class XDMSessionDetailsTests: XCTestCase {
+        }
+        """
+
+        assertExactMatch(expected: expected, actual: decodedSessionDetails)
```
Use `assertEqual`

kevinlind commented on Tests/UnitTests/XDMSessionDetailsTests.swift:
```diff
@@ -94,28 +100,33 @@ class XDMSessionDetailsTests: XCTestCase {
+        }
+        """
+
+        assertExactMatch(expected: expected, actual: decodedSessionDetails)
```
Use `assertEqual`

kevinlind commented:
@timkimadobe , regarding your question on using `isEqual` instead of the custom Double.isAlmostEqual in `MediaPublicTrackerTests`, are you seeing any test failures using `isEqual`?
It appears these tests pass in the exact values expected, so there shouldn't be any difference between the actual and expected values. The `isAlmostEqual` logic might be overkill in this case.

timkimadobe committed:
Exclusively use assertEqual for increased test strictness

This requires test case changes in XDMMediaCollectionTests.swift to reflect the two missing properties in the original test case compared to the full dictionary

timkimadobe reviewed (commented):
Thanks for the review @kevinlind! Updated to use `assertEqual` across the board, with the `XDMMediaCollectionTests` -> `testEncode_sessionStart()` requiring two additional properties added to the expected JSON as they were not present in the original test case
